File: src/main/java/com/rvbenlg/adventofcode/year2015/Day06.java
package com.rvbenlg.adventofcode.year2015;

import com.rvbenlg.adventofcode.utils.Utilities;

import java.io.IOException;
import java.util.List;

public class Day06 {

    /*
    --- Day 6: Probably a Fire Hazard ---
    Because your neighbors keep defeating you in the holiday house decorating contest year after year, you've decided to deploy one million lights in a 1000x1000 grid.

    Furthermore, because you've been especially nice this year, Santa has mailed you instructions on how to display the ideal lighting configuration.

    Lights in your grid are numbered from 0 to 999 in each direction; the lights at each corner are at 0,0, 0,999, 999,999, and 999,0. The instructions include whether to turn on, turn off, or toggle various inclusive ranges given as coordinate pairs. Each coordinate pair represents opposite corners of a rectangle, inclusive; a coordinate pair like 0,0 through 2,2 therefore refers to 9 lights in a 3x3 square. The lights all start turned off.

    To defeat your neighbors this year, all you have to do is set up your lights by doing the instructions Santa sent you in order.

    For example:

    turn on 0,0 through 999,999 would turn on (or leave on) every light.
    toggle 0,0 through 999,0 would toggle the first line of 1000 lights, turning off the ones that were on, and turning on the ones that were off.
    turn off 499,499 through 500,500 would turn off (or leave off) the middle four lights.
    After following the instructions, how many lights are lit?

    --- Part Two ---
    You just finish implementing your winning light pattern when you realize you mistranslated Santa's message from Ancient Nordic Elvish.

    The light grid you bought actually has individual brightness controls; each light can have a brightness of zero or more. The lights all start at zero.

    The phrase turn on actually means that you should increase the brightness of those lights by 1.

    The phrase turn off actually means that you should decrease the brightness of those lights by 1, to a minimum of zero.

    The phrase toggle actually means that you should increase the brightness of those lights by 2.

    What is the total brightness of all lights combined after following Santa's instructions?

    For example:

    turn on 0,0 through 0,0 would increase the total brightness by 1.
    toggle 0,0 through 999,999 would increase the total brightness by 2000000.
     */

    private boolean[][] lights = new boolean[1000][1000];
    private int[][] brightness = new int[1000][1000];

    public void solve() throws IOException {
        part1();
        part2();
    }

    private void part1() throws IOException {
        List<String> input = Utilities.readInput("year2015/day06.txt");
        input.replaceAll(instruction -> instruction.replaceAll("turn ", ""));
        followInstructions(input);
        int result = howManyLightsAreLit();
        System.out.println("Part 1 solution: " + result);
    }

    private void part2() throws IOException {
        List<String> input = Utilities.readInput("year2015/day06.txt");
        input.replaceAll(instruction -> instruction.replaceAll("turn ", ""));
        followInstructions2(input);
        int result = getTotalBrightness();
        System.out.println("Part 2 solution: " + result);
    }

    private int getTotalBrightness() {
        int result = 0;
        for (int i = 0; i < brightness.length; i++) {
            for (int j = 0; j < brightness[i].length; j++) {
                result += brightness[i][j];
            }
        }
        return result;
    }

    private void followInstructions2(List<String> input) {
        for (String instruction : input) {
            String[] words = instruction.split(" ");
            String[] from = words[1].split(",");
            String[] to = words[3].split(",");
            int fromX = Integer.parseInt(from[0]);
            int fromY = Integer.parseInt(from[1]);
            int toX = Integer.parseInt(to[0]);
            int toY = Integer.parseInt(to[1]);
            switch (words[0]) {
                case "on":
                    turnOn2(fromX, fromY, toX, toY);
                    break;
                case "off":
                    turnOff2(fromX, fromY, toX, toY);
                    break;
                case "toggle":
                    toggle2(fromX, fromY, toX, toY);
                    break;
            }
        }
    }

    private void turnOn2(int fromX, int fromY, int toX, int toY) {
        for (int i = fromY; i <= toY; i++) {
            for (int j = fromX; j <= toX; j++) {
                brightness[i][j] = brightness[i][j] + 1;
            }
        }
    }

    private void turnOff2(int fromX, int fromY, int toX, int toY) {
        for (int i = fromY; i <= toY; i++) {
            for (int j = fromX; j <= toX; j++) {
                if (brightness[i][j] > 0) {
                    brightness[i][j] = brightness[i][j] - 1;
                }
            }
        }
    }

    private void toggle2(int fromX, int fromY, int toX, int toY) {
        for (int i = fromY; i <= toY; i++) {
            for (int j = fromX; j <= toX; j++) {
                brightness[i][j] = brightness[i][j] + 2;
            }
        }
    }

    private int howManyLightsAreLit() {
        int result = 0;
        for (int i = 0; i < lights.length; i++) {
            for (int j = 0; j < lights[i].length; j++) {
                if (lights[i][j]) {
                    result++;
                }
            }
        }
        return result;
    }

    private void followInstructions(List<String> input) {
        for (String instruction : input) {
            String[] words = instruction.split(" ");
            String[] from = words[1].split(",");
            String[] to = words[3].split(",");
            int fromX = Integer.parseInt(from[0]);
            int fromY = Integer.parseInt(from[1]);
            int toX = Integer.parseInt(to[0]);
            int toY = Integer.parseInt(to[1]);
            switch (words[0]) {
                case "on":
                    turnOn(fromX, fromY, toX, toY);
                    break;
                case "off":
                    turnOff(fromX, fromY, toX, toY);
                    break;
                case "toggle":
                    toggle(fromX, fromY, toX, toY);
                    break;
            }
        }
    }

    private void turnOn(int fromX, int fromY, int toX, int toY) {
        for (int i = fromY; i <= toY; i++) {
            for (int j = fromX; j <= toX; j++) {
                lights[i][j] = true;
            }
        }
    }

    private void turnOff(int fromX, int fromY, int toX, int toY) {
        for (int i = fromY; i <= toY; i++) {
            for (int j = fromX; j <= toX; j++) {
                lights[i][j] = false;
            }
        }
    }

    private void toggle(int fromX, int fromY, int toX, int toY) {
        for (int i = fromY; i <= toY; i++) {
            for (int j = fromX; j <= toX; j++) {
                lights[i][j] = !lights[i][j];
            }
        }
    }

    private void resetVariables() {
        lights = new boolean[1000][1000];
        brightness = new int[1000][1000];
    }

}

File: year2015/solutions/day06.py
class Part1:
    """
    --- Day 6: Probably a Fire Hazard ---
    Because your neighbors keep defeating you in the holiday house decorating contest year after year, you've decided to deploy one million lights in a 1000x1000 grid.

    Furthermore, because you've been especially nice this year, Santa has mailed you instructions on how to display the ideal lighting configuration.

    Lights in your grid are numbered from 0 to 999 in each direction; the lights at each corner are at 0,0, 0,999, 999,999, and 999,0. The instructions include whether to turn on, turn off, or toggle various inclusive ranges given as coordinate pairs. Each coordinate pair represents opposite corners of a rectangle, inclusive; a coordinate pair like 0,0 through 2,2 therefore refers to 9 lights in a 3x3 square. The lights all start turned off.

    To defeat your neighbors this year, all you have to do is set up your lights by doing the instructions Santa sent you in order.

    For example:

    turn on 0,0 through 999,999 would turn on (or leave on) every light.
    toggle 0,0 through 999,0 would toggle the first line of 1000 lights, turning off the ones that were on, and turning on the ones that were off.
    turn off 499,499 through 500,500 would turn off (or leave off) the middle four lights.
    After following the instructions, how many lights are lit?
    """

    def __init__(self, input: list[str] = None) -> None:
        self.input = input

    def solve(self) -> int:
        """Solver for part 1 of the puzzle.

        Returns:
            int: Solution of part 1
        """
        return self.how_many_lit_lights(self.input)

    def how_many_lit_lights(self, input: list[str]) -> int:
        """Count how many lights are lit.

        Args:
            input (list[str]): List of instructions

        Returns:
            int: Number of lit lights
        """
        lights = self.initialize_lights()
        for instruction in input:
            lights = self.execute_instruction(lights, instruction)
        return self.count_lit_lights(lights)

    def initialize_lights(self) -> list[list[bool]]:
        """Initialize lights.

        Returns:
            list[list[bool]]: Lights
        """
        lights = []
        for _ in range(1000):
            lights.append([False] * 1000)
        return lights

    def execute_instruction(
        self, lights: list[list[bool]], instruction: str
    ) -> list[list[bool]]:
        """Execute an instruction.

        Args:
            lights (list[list[bool]]): Lights
            instruction (str): Instruction

        Returns:
            list[list[bool]]: Lights
        """
        instruction = instruction.split(" ")
        if instruction[0] == "toggle":
            return self.toggle_lights(lights, instruction)
        elif instruction[0] == "turn":
            return self.turn_lights(lights, instruction)
        else:
            raise ValueError(f"Unknown instruction: {instruction}")

    def toggle_lights(
        self, lights: list[list[bool]], instruction: list[str]
    ) -> list[list[bool]]:
        """Toggle lights.

        Args:
            lights (list[list[bool]]): Lights
            instruction (list[str]): Instruction

        Returns:
            list[list[bool]]: Lights
        """
        start = instruction[1].split(",")
        end = instruction[3].split(",")
        for i in range(int(start[0]), int(end[0]) + 1):
            for j in range(int(start[1]), int(end[1]) + 1):
                lights[i][j] = not lights[i][j]
        return lights

    def turn_lights(
        self, lights: list[list[bool]], instruction: list[str]
    ) -> list[list[bool]]:
        """Turn lights on or off.

        Args:
            lights (list[list[bool]]): Lights
            instruction (list[str]): Instruction

        Returns:
            list[list[bool]]: Lights
        """
        start = instruction[2].split(",")
        end = instruction[4].split(",")
        for i in range(int(start[0]), int(end[0]) + 1):
            for j in range(int(start[1]), int(end[1]) + 1):
                if instruction[1] == "on":
                    lights[i][j] = True
                elif instruction[1] == "off":
                    lights[i][j] = False
                else:
                    raise ValueError(f"Unknown instruction: {instruction}")
        return lights

    def count_lit_lights(self, lights: list[list[bool]]) -> int:
        """Count how many lights are lit.

        Args:
            lights (list[list[bool]]): Lights

        Returns:
            int: Number of lit lights
        """
        lit_lights = 0
        for i in range(len(lights)):
            for j in range(len(lights[i])):
                if lights[i][j]:
                    lit_lights += 1
        return lit_lights


class Part2:
    """
    --- Part Two ---
    You just finish implementing your winning light pattern when you realize you mistranslated Santa's message from Ancient Nordic Elvish.

    The light grid you bought actually has individual brightness controls; each light can have a brightness of zero or more. The lights all start at zero.

    The phrase turn on actually means that you should increase the brightness of those lights by 1.

    The phrase turn off actually means that you should decrease the brightness of those lights by 1, to a minimum of zero.

    The phrase toggle actually means that you should increase the brightness of those lights by 2.

    What is the total brightness of all lights combined after following Santa's instructions?

    For example:

    turn on 0,0 through 0,0 would increase the total brightness by 1.
    toggle 0,0 through 999,999 would increase the total brightness by 2000000.
    """

    def __init__(self, input: list[str] = None) -> None:
        self.input = input

    def solve(self) -> int:
        """Solver for part 2 of the puzzle.

        Returns:
            int: Solution of part 2
        """
        return self.calculate_total_brightness(self.input)

    def calculate_total_brightness(self, input: list[str]) -> int:
        """Calculate total brightness.

        Args:
            input (list[str]): List of instructions

        Returns:
            int: Total brightness
        """
        lights = self.initialize_lights()
        for instruction in input:
            lights = self.execute_instruction(lights, instruction)
        return self.calculate_total_brightness_of_lights(lights)

    def initialize_lights(self) -> list[list[int]]:
        """Initialize lights.

        Returns:
            list[list[int]]: Lights
        """
        lights = []
        for _ in range(1000):
            lights.append([0] * 1000)
        return lights

    def execute_instruction(
        self, lights: list[list[int]], instruction: str
    ) -> list[list[int]]:
        """Execute an instruction.

        Args:
            lights (list[list[int]]): Lights
            instruction (str): Instruction

        Returns:
            list[list[int]]: Lights
        """
        instruction = instruction.split(" ")
        if instruction[0] == "toggle":
            return self.toggle_lights(lights, instruction)
        elif instruction[0] == "turn":
            return self.turn_lights(lights, instruction)
        else:
            raise ValueError(f"Unknown instruction: {instruction}")

    def toggle_lights(
        self, lights: list[list[int]], instruction: str
    ) -> list[list[int]]:
        """Toggle lights.

        Args:
            lights (list[list[int]]): Lights
            instruction (str): Instruction

        Returns:
            list[list[int]]: Lights
        """
        start = instruction[1].split(",")
        end = instruction[3].split(",")
        for i in range(int(start[0]), int(end[0]) + 1):
            for j in range(int(start[1]), int(end[1]) + 1):
                lights[i][j] += 2
        return lights

    def turn_lights(self, lights: list[list[int]], instruction: str) -> list[list[int]]:
        """Turn lights on or off.

        Args:
            lights (list[list[int]]): Lights
            instruction (str): Instruction

        Returns:
            list[list[int]]: Lights
        """
        start = instruction[2].split(",")
        end = instruction[4].split(",")
        for i in range(int(start[0]), int(end[0]) + 1):
            for j in range(int(start[1]), int(end[1]) + 1):
                if instruction[1] == "on":
                    lights[i][j] += 1
                elif instruction[1] == "off":
                    if lights[i][j] > 0:
                        lights[i][j] -= 1
                else:
                    raise ValueError(f"Unknown instruction: {instruction}")
        return lights

    def calculate_total_brightness_of_lights(self, lights: list[list[int]]) -> int:
        """Calculate total brightness of lights.

        Args:
            lights (list[list[int]]): Lights

        Returns:
            int: Total brightness of lights
        """
        total_brightness = 0
        for i in range(len(lights)):
            for j in range(len(lights[i])):
                total_brightness += lights[i][j]
        return total_brightness
